crates/pixi_record/src/source_record.rs

```diff
@@ -55,80 +64,129 @@ pub struct InputHash {
-            PinnedSourceSpec::Url(pinned_url_spec) => PackageBuildSource::Url {
-                url: pinned_url_spec.url,
-                sha256: pinned_url_spec.sha256,
-                subdir: None,
-            },
-            PinnedSourceSpec::Git(pinned_git_spec) => {
-                let subdirectory = pinned_git_spec
-                    .source
-                    .subdirectory
-                    .as_deref()
-                    .map(Utf8TypedPathBuf::from);
-
-                let spec = match &pinned_git_spec.source.reference {
-                    GitReference::Branch(branch) => Some(GitShallowSpec::Branch(branch.clone())),
-                    GitReference::Tag(tag) => Some(GitShallowSpec::Tag(tag.clone())),
-                    GitReference::Rev(_) => Some(GitShallowSpec::Rev),
-                    GitReference::DefaultBranch => None,
-                };
-
-                PackageBuildSource::Git {
-                    url: pinned_git_spec.git,
-                    spec,
-                    rev: pinned_git_spec.source.commit.to_string(),
-                    subdir: subdirectory,
-                }
-            }
-            PinnedSourceSpec::Path(pinned_path) => PackageBuildSource::Path {
-                path: pinned_path.path,
-            },
-        });
-        CondaPackageData::Source(CondaSourceData {
-            package_record: value.package_record,
-            location: value.manifest_source.clone().into(),
+impl SourceRecord {
+    /// Convert [`SourceRecord`] into lock-file compatible `CondaSourceData`
+    /// The `build_source` in the SourceRecord is always relative to the workspace.
+    /// However, when saving in the lock-file make these relative to the package manifest.
+    /// This should be used when writing to the lock file.
+    pub fn into_conda_source_data(self, workspace_root: &Path) -> CondaSourceData {
+        let package_build_source =
+            self.build_source
+                .clone()
+                .map(|build_source| match build_source {
+                    PinnedSourceSpec::Git(git_spec) => {
+                        // When manifest and build refer to the same repo+commit we keep the
+                        // build source as git, but rewrite its subdirectory relative to the
+                        // manifest checkout as expected by the lock file format.
+                        let mut subdirectory = git_spec.source.subdirectory.clone();
+
+                        if let PinnedSourceSpec::Git(manifest_git) = &self.manifest_source {
+                            if same_git_checkout(&git_spec, manifest_git) {
+                                subdirectory = relative_repo_subdir(
+                                    manifest_git.source.subdirectory.as_deref().unwrap_or(""),
+                                    git_spec.source.subdirectory.as_deref().unwrap_or(""),
+                                );
+                            }
+                        }
```

Comment:
This is not needed. A git reference is already "absolute", we can keep it that way.